Add a launcher for the gateway.
--- /dev/null
+++ b/webthings-gateway
@@ -0,0 +1,16 @@
+#!/bin/sh
+
+if ! [ $(id -u) = 0 ]; then
+  echo "Program must be run as root."
+  exit 1
+fi
+
+if [ ! -d "/var/lib/webthings-gateway" ]; then
+  mkdir "/var/lib/webthings-gateway"
+  chown -R webthings:webthings "/var/lib/webthings-gateway"
+fi
+
+cd /usr/lib/webthings-gateway
+su -c "node build/gateway.js" webthings
+
+# vim:set ts=2 sw=2 et ft=sh:
